src/Temporalio/Worker/ActivityWorker.cs

```diff
@@ -279,32 +292,49 @@ private void StartActivity(Bridge.Api.ActivityTask.ActivityTask tsk)
                 // Find activity or fail
                 if (!activities.TryGetValue(act.Context.Info.ActivityType, out var defn))
                 {
-                    var avail = activities.Keys.ToList();
-                    avail.Sort();
-                    var availStr = string.Join(", ", avail);
-                    throw new ApplicationFailureException(
-                        $"Activity {act.Context.Info.ActivityType} is not registered on this worker," +
-                        $" available activities: {availStr}",
-                        errorType: "NotFoundError");
+                    defn = dynamicActivity;
+                    if (defn == null)
+                    {
+                        var avail = activities.Keys.ToList();
+                        avail.Sort();
+                        var availStr = string.Join(", ", avail);
+                        throw new ApplicationFailureException(
+                            $"Activity {act.Context.Info.ActivityType} is not registered on this worker," +
+                            $" available activities: {availStr}",
```

Comment:
Ug, yeah that's the size of the entire failure proto - message, stack trace, recursive causes and all. What should we do for big stack traces? What should we do for large cause chains?

Seems it's not listing activities is problematic, but large failures are problematic and I wonder if we need a holistic solution. How am I supposed to know whether I can list more or less activities just because the stack trace happens to be smaller?

Reply:
Indeed, the problem is not specifically with listing activity names. Yet, the fact that the server may penalize us on for sending the list of activity names if that exceeds some threshold means that this is not a reliable way to provide that kind of feedback to user. We'll need to find an alternative for listing registered activity/workflow/signal/queries types AND a better way to deal with activities failures containing large stack traces.

Also, I'd say that Failure should not contain stack trace for these kind of errors, if that's possible. There is no user code on the stack at this point, so nothing they should need to trace.

But really, that don't need to be resolved today.

Reply:
> server may penalize us on for sending the list of activity names if that exceeds some threshold

Yeah, I need to probably put some upper bound on this

> We'll need to find an alternative for listing registered activity/workflow/signal/queries types 

https://github.com/temporalio/features/issues/51

> I'd say that Failure should not contain stack trace for these kind of errors, if that's possible

It's possible, but not very transparent. As an open source library I'm like showing where all exceptions occur even in our own code.